File: day6/day6.cpp
#include <algorithm>
#include <fstream>
#include <fmt/format.h>

using namespace std;

static pair<int, int> sum_common_answers(istream& input)
{
	int any = 0;
	int all = 0;
	int group_count = 0;
	int answers[26];
	fill(answers, answers+26, 0);

	string line;
	do
	{
		if (getline(input, line) && line.size())
		{
			for (auto letter: line)
			{
				if ('a' <= letter && letter <= 'z')
				{
					answers[letter-'a']++;
				}
			}
			group_count++;
		}
		else if (group_count)
		{
			for (auto a: answers)
			{
				any += a > 0;
				all += a == group_count;
			}
			fill(answers, answers+26, 0);
			group_count = 0;
		}
	} while (input);

	return make_pair(any, all);
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fmt::print(stderr, "Usage: {} <filename>\n", argv[0]);
		return 1;
	}

	ifstream input(argv[1]);
	if (!input)
	{
		fmt::print(stderr, "Cannot open {}\n", argv[1]);
		return 1;
	}

	auto [part1, part2] = sum_common_answers(input);
	input.close();

	fmt::print("Part1: {}\nPart2: {}\n", part1, part2);
	return 0;
}

File: day6/day6.py
#!/usr/bin/env python3

import sys

def sum_common_answers(txt):
    sum_any = 0
    sum_all = 0
    answers = [0] * 26
    group_count = 0

    for line in txt.split("\n"):
        line = line.strip()
        if line:
            for letter in line:
                if ord('a') <= ord(letter) <= ord('z'):
                    answers[ord(letter) - ord('a')] += 1
            group_count += 1
        elif group_count:
            for answer in answers:
                if answer > 0:
                    sum_any += 1
                if answer == group_count:
                    sum_all += 1

            answers = [0] * 26
            group_count = 0

    if group_count:
        for answer in answers:
            if answer > 0:
                sum_any += 1
            if answer == group_count:
                sum_all += 1

    return sum_any, sum_all

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read()
    except:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)

    part1, part2 = sum_common_answers(txt)
    print("Part1:", part1)
    print("Part2:", part2)
